### ISSUE TITLE
[TODO] Get rid of unittest.TestCase

### ISSUE THREAD

**Issue by webknjaz** [cherrypy/cheroot#42]
This is needed to fully take advantage of `py.test` testing framework.

**Comment by webknjaz** [comment]
Ref: https://stackoverflow.com/a/35562401/595220
Ref: https://docs.pytest.org/en/latest/unittest.html

**Comment by jaraco** [comment]
Fully support.

**Comment by jaraco** [comment]
One consideration is some environments may be relying on WebCase as a unittest.TestCase subclass. I suggest separating the web testing functionality into a WebTest class that can be instantiated without any unittest functionality, but also use it as a mix-in for WebCase to continue to provide that functionality in the WebCase.

**Comment by webknjaz** [comment]
I've dropped a few classes in #67

**Comment by webknjaz** [comment]
Next thing to achieve on this is to deprecate webtest and try to make some public test fixture/helper module

**Comment by jaraco** [comment]
In my experience - CherryPy's webtest is helpful, but it also masks errors (because every request happens on a remote side of a call, it's impossible to debug with something like pdb). I've found instead the [webtest](https://pypi.org/project/WebTest/) project and more interestingly [pyriform](https://pypi.org/project/pyriform/), though admittedly I haven't used the latter yet.

**Comment by webknjaz** [comment]
I was also thinking of pylons.webtest and officially recommending to defer WSGI testing to third party, but didn't explore the possibilities and effect of such decision.

**Comment by webknjaz** [comment]
So I looked through pyriform and it looks simple and easy to use. There's no pytest fixtures shipped though, which might be useful. On the other side, we could provide our own fixtures based on this library. I foresee that it would fit in our ecosystem :)

**Comment by jaraco** [comment]
I don't think there is anything called [pylons.webtest](https://pypi.org/project/pylons.webtest/) and when I find [Pylons/webtest](/Pylons/webtest), it seems to be one and the same project as the one I referenced.

And pyriform builds on that same framework.

I know the developer of pyriform and I bet he'd accept pull requests for pytest fixtures, so something definitely worth considering contributing upstream.

**Comment by webknjaz** [comment]
unstale

**Comment by webknjaz** [comment]
unstale

**Comment by jeffvanvoorst** [comment]
I would like to tackle this.  I am familiar with the requests, pytest, and unittest modules.

I did see that pyriform links the requests and WebTest modules.  Is it OK to have all 3 modules and dependencies as requirements for unit tests?

My reason for asking is that I was reading a CherryPy PR with a comment that "we probably don't want to introduce this dependency right now".  https://github.com/cherrypy/cherrypy/pull/1715.

**Comment by webknjaz** [comment]
@jeffvanvoorst it'd be nice to get some help!

I've done some refactoring introducing some replacements for server creation and client generation as fixtures (as opposed to https://github.com/cherrypy/cheroot/blob/master/cheroot/test/helper.py#L35 and https://github.com/cherrypy/cheroot/blob/master/cheroot/test/webtest.py#L99, which we're trying to make easily deprecatable in future).
Also, I've tried introducing some SSL tests (`test_ssl.py`), which *temporarily* use requests directly + there's a contribution from @dpeschman (`test_https.py`) where he modified that old client to support SSL client cert and introduced `ddt` dependency (`pytest.mark.parametrized` is not compatible with old helpers)

So we want to:
- remove `unittest` completely
- stop using old helper testcase class bases
- start using `pyriform` directly, it's okay to use it, but not put requests/webtest as a direct dependency
- stop using `requests` directly (in favor of `pyriform`) for testing WSGI
- create a better framework for testing (as a replacement of old helpers), I've got some helpers in `cheroot.testing` and `cheroot.test.conftest`, and fixtures in test modules themselves, but there's a load of copy-paste, which I desire to reduce and make more elegant

N.B. I don't want to use pure requests here, because we implement HTTP parser and low-level protocol, so we need to test (1) malicious requests, (2) tune some low-level stuff for testing, which requests being a very high-level tool don't quite facilitate.

**Comment by jeffvanvoorst** [comment]
I am thinking that given my time, it would be best to tackle one thing at a time.  For example, it seems reasonable to try to remove unittest from WebCase as a first step.  Then can continue to complete items.

**Comment by webknjaz** [comment]
@jeffvanvoorst do it! it comes from https://github.com/cherrypy/cheroot/pull/52/files#r213702146
I think it might be possible to also use trustme for certificate generation and setting up CA into SSLContext in that https module.

**Comment by jeffvanvoorst** [comment]
@webknjaz it is taking me some time to wrap my head around the pieces.  My plan keeps morphing.  I think that I finally grokked how to use pyriform -- pass in the actual webserver instead of wsgi app.  I will try some PoCs when I get time.  After that, I should be able to more quickly code up a pytest implementation (leveraging pyriform) to test certificate handling.  If nothing else, pyriform knowledge seems to be useful -- it would be great to unit/component test some custom requests code + wsgi app for a different project.

**Comment by webknjaz** [comment]
Yeah, I see... Maybe it's easier to start with refactoring other test modules, which already use pytest, to create a foundation for further refactoring?

**Comment by webknjaz** [comment]
unstale
